storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java

```diff
@@ -0,0 +1,222 @@
+        }
+        return null;
+    }
+
+    @Override
+    public Set<SubSystem> getSubSystems() {
+
+        Set<SubSystem> subSystems = new HashSet<SubSystem>();
+
+        try (FileReader reader = new FileReader(CgroupUtils.CGROUP_STATUS_FILE);
+             BufferedReader br = new BufferedReader(reader)){
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] split = str.split("\t");
+                SubSystemType type = SubSystemType.getSubSystem(split[0]);
+                if (type == null) {
+                    continue;
+                }
+                subSystems.add(new SubSystem(type, Integer.valueOf(split[1]), Integer.valueOf(split[2]),
+                        Integer.valueOf(split[3]).intValue() == 1 ? true : false));
+            }
+            return subSystems;
+        } catch (Exception e) {
+            LOG.error("Get subSystems error {}", e);
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isSubSystemEnabled(SubSystemType subSystemType) {
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
+            if (subSystem.getType() == subSystemType) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Hierarchy getHierarchyWithSubSystem(SubSystemType subSystem) {
+        return getHierarchyWithSubSystems(Arrays.asList(subSystem));
+    }
+
+    @Override
+    public Hierarchy getHierarchyWithSubSystems(List<SubSystemType> subSystems) {
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        for (Hierarchy hierarchy : hierarchies) {
+            Hierarchy ret = hierarchy;
+            for (SubSystemType subSystem : subSystems) {
+                if (!hierarchy.getSubSystems().contains(subSystem)) {
+                    ret = null;
+                    break;
+                }
+            }
+            if (ret != null) {
+                return ret;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isMounted(Hierarchy hierarchy) {
+
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        if (Utils.CheckDirExists(hierarchy.getDir())) {
+            for (Hierarchy h : hierarchies) {
+                if (h.equals(hierarchy)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public void mount(Hierarchy hierarchy) throws IOException {
+
+        if (this.isMounted(hierarchy)) {
+            LOG.error("{} is already mounted", hierarchy.getDir());
+            return;
+        }
+        Set<SubSystemType> subSystems = hierarchy.getSubSystems();
+        for (SubSystemType type : subSystems) {
+            if (this.getHierarchyWithSubSystem(type) != null) {
+                LOG.error("subSystem: {} is busy", type.name());
+                subSystems.remove(type);
```

Comment:
It modifies the contents of hierarchy object also which can have side effects. Do you want to duplicate subSystems?